File: 261.graph-valid-tree.cpp
/*
 * @lc app=leetcode id=261 lang=cpp
 *
 * [261] Graph Valid Tree
 */
#include <vector>

using std::vector;

// @lc code=start
class Solution {
public:
    // Solution based on Union-Find
    // T: O(n * a(n)) ~= O(n), a(n) := Inverse Ackermann Function (grows slowly, can be treated as a constant)
    // S: O(n)
    bool validTree(int n, vector<vector<int>>& edges) {
        // Check if any nodes not connected or if there exist duplicate edge(s)
        if (edges.size() != (n - 1))  return false;

        vector<int> nodes(n, 0);
        for (int i = 0; i < n; ++i)  nodes[i] = i;
        for (const auto &e : edges) {
            int v1 = e[0], v2 = e[1];
            while (nodes[v1] != v1) {
                nodes[v1] = nodes[nodes[v1]];   // path compression
                v1 = nodes[v1];
            }
            while (nodes[v2] != v2) { 
                nodes[v2] = nodes[nodes[v2]];   // path compression
                v2 = nodes[v2];
            }
            if (nodes[v1] == nodes[v2])  return false;  // cycle found
            nodes[v2] = v1;     // might consider rank/weight/size here
        }
        return true;
    }
};
// @lc code=end

File: 261.graph-valid-tree.py
#
# @lc app=leetcode id=261 lang=python3
#
# [261] Graph Valid Tree
#
from typing import *

# @lc code=start
class UF:
    # S: O(n)
    def __init__(self, n: int):
        self.cnt = n
        self.parents = list(range(n))

        # Store how many nodes are connected to each node, 
        # which is used in connecting two nodes to avoid 
        # degradation to linked list
        self.sizes = [1] * n

    # With rebalancing, T is O(log(n))
    def union(self, p: int, q: int) -> bool:
        root_p, root_q = self.find(p), self.find(q)
        if root_p == root_q:
            # Already connected
            return False

        # Rebalancing, avoid degradation to linked list
        # connect smaller subgraph to the larger one
        if self.sizes[root_p] > self.sizes[root_q]:
            self.parents[root_q] = root_p
            self.sizes[root_p] += self.sizes[root_q]
        else:
            self.parents[root_p] = root_q
            self.sizes[root_q] += self.sizes[root_p]

        self.cnt -= 1
        return True

    # With path compression, T can be reduced to O(1) eventually
    def find(self, p: int) -> int:
        while p != self.parents[p]:
            self.parents[p] = self.parents[self.parents[p]]
            p = self.parents[p]
        return p

class Solution:
    # T: O(V + E*log(E)), V := n, E := len(edges)
    # S: O(V)
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        uf = UF(n)  
        for p, q in edges:
            if not uf.union(p, q):
                # If p and q are already connected,
                # then there must exist a cycle
                return False
        return uf.cnt == 1  # check if fully connected (if no, a forest)

# @lc code=end

